torch/_overrides.py

```diff
@@ -0,0 +1,242 @@
+"""
+Preliminary implementation of __torch_function__
+
+TODO: rewrite this in C++ for performance.
+
+NOTE: heavily inspired by NumPy's ``__array_function__`` (see:
+https://github.com/pytorch/pytorch/issues/24015 and
+https://www.numpy.org/neps/nep-0018-array-function-protocol.html
+)
+
+"""
+
+import collections
+import functools
+import textwrap
+
+# TODO: PyTorch does not have a hard dependency on NumPy, so we need
+#       to vendor this code.
+from numpy.compat._inspect import getargspec
+
+from .tensor import Tensor
+
+
+_TORCH_FUNCTION = Tensor.__torch_function__
+_TENSOR_ONLY = [Tensor]
+
+def get_overloaded_types_and_args(relevant_args):
+    """Returns a list of arguments on which to call __torch_function__.
+
+    Parameters
+    ----------
+    relevant_args : iterable of array-like
+        Iterable of array-like arguments to check for __torch_function__
+        methods.
+
+    Returns
+    -------
+    overloaded_types : collection of types
+        Types of arguments from relevant_args with __torch_function__ methods.
+    overloaded_args : list
+        Arguments from relevant_args on which to call __torch_function__
+        methods, in the order in which they should be called.
+
+    """
+    # Runtime is O(num_arguments * num_unique_types)
+    overloaded_types = []
+    overloaded_args = []
+    for arg in relevant_args:
+        arg_type = type(arg)
+        # We only collect arguments if they have a unique type, which ensures
+        # reasonable performance even with a long list of possibly overloaded
+        # arguments.
+        if (arg_type not in overloaded_types and
+                hasattr(arg_type, '__torch_function__')):
+
+            # Create lists explicitly for the first type (usually the only one
+            # done) to avoid setting up the iterator for overloaded_args.
+            if overloaded_types:
+                overloaded_types.append(arg_type)
+                # By default, insert argument at the end, but if it is
+                # subclass of another argument, insert it before that argument.
+                # This ensures "subclasses before superclasses".
+                index = len(overloaded_args)
+                for i, old_arg in enumerate(overloaded_args):
+                    if issubclass(arg_type, type(old_arg)):
+                        index = i
+                        break
+                overloaded_args.insert(index, arg)
+            else:
+                overloaded_types = [arg_type]
+                overloaded_args = [arg]
+
+    return overloaded_types, overloaded_args
+
+
+def implement_torch_function(
+        implementation, public_api, relevant_args, args, kwargs):
+    """Implement a function with checks for __torch_function__ overrides.
```

Comment:
Though not a decorator, intruigingly.

Reply:
`torch_function_dispatch` is the decorator. this generates the function that that decorator returns, and is used from within `torch_function_dispatch` only.